File: 140-word-break-ii/140-word-break-ii.java
class Solution {
  public void solve(String s, int startIdx, int currIdx, String sentence, Set<String> wordDict, List<String> res)
  {
    if(currIdx == s.length())
    {
      if(startIdx < currIdx)
      {
        String currWord = s.substring(startIdx, currIdx);
        if(wordDict.contains(currWord))
        {
          if(sentence!="")
            sentence += " ";
          sentence += currWord;
        }
        else
          return;
      }
      res.add(sentence);
      return;
    }
    String currWord = s.substring(startIdx, currIdx);
    if(wordDict.contains(currWord))
    {
      String nextSentence = sentence;
      if(nextSentence!="")
        nextSentence += " ";
      nextSentence += currWord;
      solve(s, currIdx, currIdx+1, nextSentence, wordDict, res);
    }
    solve(s, startIdx, currIdx+1, sentence, wordDict, res);
  }
  
  public List<String> wordBreak(String s, List<String> wordDict) {
    List<String> res = new ArrayList();
    Set<String> mp = new HashSet(wordDict);
    solve(s, 0, 0, "", mp, res);
    return res;
  }
}

File: 140-word-break-ii/140-word-break-ii.cpp
class Solution {
public:
  
  void solve(string s, string currPath, string currWord, unordered_map<string, bool>& wordDict, vector<string>& res)
  {
    if(s == "")
    {
      if(wordDict[currWord])
      {
        currPath += currPath == "" ? currWord : " " + currWord;
        res.push_back(currPath);
      }
      return;
    }
    currWord += s[0];
    s.erase(0,1);
    if(wordDict[currWord])
    {
      solve(s, (currPath == "") ? currWord : (currPath + " " + currWord), "", wordDict, res);
    }
    solve(s, currPath, currWord, wordDict, res);
  }
  
  vector<string> wordBreak(string s, vector<string>& wordDict) {
    unordered_map<string,bool> mp;
    for(string word : wordDict)
    {
      mp[word] = true;
    }
    vector<string> res;
    solve(s, "", "", mp, res);
    return res;
  }
};